src/game_api/memory.hpp

```diff
@@ -59,10 +133,61 @@ void write_mem_prot(size_t addr, std::string payload, bool prot)
 {
     write_mem_prot(addr, payload, false);
 }
-#define DEFINE_ACCESSOR(name, type)                \
-    [[maybe_unused]] type read_##name(size_t addr) \
-    {                                              \
-        return *(type*)(addr);                     \
+
+struct ReversibleMemory
+{
+    size_t address;
+    std::vector<char> old_data;
+    bool prot_used;
+};
+
+std::unordered_map<std::string, std::vector<ReversibleMemory>> original_memory;
+
+[[maybe_unused]] void write_mem_reversible(std::string name, size_t addr, std::string_view payload, bool prot)
+{
+    if (!original_memory.contains(name))
+    {
+        std::vector<char> old_data((char*)addr, (char*)addr + payload.size());
+        original_memory[name] = std::vector<ReversibleMemory>{{addr, old_data, prot}};
+    }
+    else
+    {
+        bool new_addr = true;
+        for (auto& it : original_memory[name])
+        {
+            if (it.address == addr)
+            {
+                new_addr = false;
+                break;
+            }
+        }
+        if (new_addr)
+        {
+            std::vector<char> old_data((char*)addr, (char*)addr + payload.size());
+            original_memory[name].push_back(ReversibleMemory{addr, old_data, prot});
+        }
+    }
+    write_mem_prot(addr, payload, prot);
+}
+template <class T>
+requires(std::is_trivially_copyable_v<T> && !std::is_same_v<T, std::string_view>) void write_mem_reversible(std::string name, size_t addr, const T& payload, bool prot)

```

Comment:
Can you move the concept to a new line and wrap the whole function in `// clang-format off` and `// clang-format on` please?
This is kind of unreadable  :(